### DISCUSSION TITLE
is Teal strongly or weakly typed? Static or Dynamic?

### DISCUSSION THREAD

**Discussion by greensourcecode** [teal-language/tl#505] [Q&A] [answered]
Title says it all

**Comment by hishamhm** [comment, marked as answer]
"Statically vs dynamically" is a more well-defined distinction, that is agreed upon in academia. Teal is statically typed; Lua is dynamically typed.

"Strongly vs weakly" is often used to mean the same as the above, but sometimes it is used to mean that you can map memory to different types. In the latter sense, both are strongly typed, in the sense that you can't really map a memory location to a different type (cast in the C sense); in Teal you have a cast, but attempting to use a value as the wrong type will cause a runtime error.

**Reply by greensourcecode** [reply]
Thank you for such a thorough answer. I'm still very green which I'm sure is no surprise. The reason I'm asking is this graph:
![strongweakstaticdynamic_type.png](https://user-images.githubusercontent.com/97241812/152667338-65b12c65-4569-48a3-9bb9-ea6c7d6c8351.png)

About where would you put Teal on this?

**Reply by cody271** [reply]
Teal is statically typed, and would be considered more strongly typed than Lua itself, for example in type coercions:
```sh
$ cat strongly_typed.tl 
```
```lua
print(1 + "1")
```

```sh
$ tl run strongly_typed.tl 
========================================
1 error:
strongly_typed.tl:1:9: cannot use operator '+' for types integer and string "1"
```

```sh
$ lua strongly_typed.tl 
2
```

**Reply by hishamhm** [reply]
Not sure about that graph, because everyone has a different definition of what "strong vs weak" means, and I can't even tell what's the desired definition from the graph alone.

My recommendation in general is to use statically vs dynamically typed to when you want to say that the variables have types or just the values, and to avoid the strongly vs weakly typed confusion altogether.

**Reply by cody271** [reply]
> Not sure about that graph, because everyone has a different definition of what "strong vs weak" means, and I can't even tell what's the desired definition from the graph alone.
>
Exactly. "Static vs dynamic" means something in computer science terms, but "strong vs weak" is poorly defined and somewhat ambiguous.

If we stick with the specific example of type coercions: Python and Ruby would produce a similar error as Teal for `1 + "1"`, but JavaScript would coerce like Lua. (Except now its `11` instead of `2`. 🤣 )

Although that's just one specific aspect. Different sources would probably draw that graph there differently..

**Reply by greensourcecode** [reply]
I see. You guys are awesome, I can't tell you how much I appreciate the knowledge. I'm actually on the spectrum and obsess over this kind of stuff. I've never done anything professionally but in the last year I've written / messed around with / learned about HTML CSS JS (starting to learn typescript), Python, Lua, C, and I've tried Rust (I'll let you guess how that went).

I like the safety of Rust and that it compiles to machine code but the lack of garbage collection creates a level of complexity that doesn't seem ideal for every project. I like the simplicity of Lua and Python, but don't like unnecessary errors. Not a HUGE fan of JavaScript. I don't HATE it I just wish there were something better. 

My ideal language (in my own naive / noob mind) would be a statically** typed and safe language that abstracts unnecessary things (like garbage collection), prevents unnecessary errors, compiles into machine code AND lua (so it can be compiled into lua JIT for browsers and maybe quick previews of the code?). Maybe tl + mlua?

Can I donate to this project (I can do monthly). I'm excited about this now!!

Also I've been told I'm the guy that calls people in the phone book to shout about dinosaurs, so if my thought process / the questions I ask are abrasive I apologize. . .

edit: I meant statically lol

**Reply by lenscas** [reply]
> and I've tried Rust (I'll let you guess how that went).

Don't worry if you didn't succeed at Rust at first. It is not the easiest language to grasp :) Slow and steady wins that race.

> compiles into machine code

Even if a language compiles directly to machine code that doesn't mean it is fast. And having a GC also means that it is also less ideal to do FFI with compared to C and Rust.

> AND lua (so it can be compiled into lua JIT for browsers and maybe quick previews of the code?)

compiling to luajit doesn't really help for it running in the browser. (In fact, it might even hurt it compared to regular lua?). Compiling directly to WASM would make it A LOT easier to run in the browser. However, as a language like you describe would most likely require quite a big runtime (the GC for example) this also becomes less ideal.  (Last time I checked, a simple hello world from some Go lang to wasm project took multiple seconds to load, presumably because Go needs a runtime that needs to get loaded as well.)

Compiling to JS could help, F# with fable seems to do quite well with this.

>  Maybe tl + mlua?
mlua, as in https://docs.rs/mlua/latest/mlua/ ? Not entirely sure how that gets close to the language features you want from an ideal language but... you are in luck as I've been working on <https://docs.rs/tealr/latest/tealr/> to make mlua + teal A LOT nicer :)

> Also I've been told I'm the guy that calls people in the phone book to shout about dinosaurs, so if my thought process / the questions I ask are abrasive I apologize. . .

So far, I don't see a problem :)

**Reply by greensourcecode** [reply]
Here is what I think is an EXCELLENT article on what we spoke about earlier:
https://ericlippert.com/2012/10/15/is-c-a-strongly-typed-or-a-weakly-typed-language/

>  Don't worry if you didn't succeed at Rust at first. It is not the easiest language to grasp :) Slow 
and steady wins that race.

I LOVE RUST. Preaching to the choir. When you're playing an RPG you can't just skip to the last boss; you have to get all the crystals first. I can't wait until I'm proficient enough to truly understand that language.

> compiling to luajit doesn't really help for it running in the browser. (In fact, it might even hurt it compared to regular lua?). Compiling directly to WASM would make it A LOT easier to run in the browser. However, as a language like you describe would most likely require quite a big runtime (the GC for example) this also becomes less ideal. (Last time I checked, a simple hello world from some Go lang to wasm project took multiple seconds to load, presumably because Go needs a runtime that needs to get loaded as well.)

Good to know, thank you for the clarification!

> However, as a language like you describe would most likely require quite a big runtime

As far as I can tell, the language I'm describing is Teal, I just didn't understand that when I originally posted. This is kind of hard to explain, but I'm going to draw out kind of what I was thinking. I will upload a picture in a little while.

Thank you guys again, I try not to ask too many questions (I think the arch guys scared me haha). This might be the single greatest conversation I've ever had though..

**Reply by lenscas** [reply]
teal doesn't compile to machine code though?

> Thank you guys again, I try not to ask too many questions (I think the arch guys scared me haha). This might be the single greatest conversation I've ever had though..

Ask away, but.. maybe on gitter instead of this thread? :sweat_smile:

**Reply by greensourcecode** [reply]
> teal doesn't compile to machine code though?

From my understanding Teal compiles to Lua which does compile into machine code? I drew a picture to help conceptualize this lol.

![1644161627390](https://user-images.githubusercontent.com/97241812/152688422-5bda6e14-7d59-4d98-86e5-c2d75a4939a0.jpg)

> Ask away, but.. maybe on gitter instead of this thread? 😅

GOOD TO KNOW. For what it's worth the society in reality took me a while to figure out too. I suppose next time I have a question I will ask there (just now learning of its existence so thank you)

**Reply by lenscas** [reply]
lua gets interpreted by a VM

mlua/rlua (assuming you refer to https://docs.rs/mlua/latest/mlua/ ) is just bindings for Rust to lua VM. It allows running lua code because not because it compiles it, but because it contains the VM and instructs that to run it. 

There are a couple of projects to get lua working on the web, but those work by compiling the lua vm to WASM or JS, and run the scripts that way. (this is also how C# runs on WASM btw)

It could be that there is a project out there that compiles lua to machine code, and LuaJIT gets close at it JIT's it but that still has a VM to my knowledge and only does this compilation for functions that it thinks are worth it and does this during runtime. Mlua and Rlua are however not these compilers. They just allow Rust code to interact with various Lua vm's.

**Reply by catwell** [reply]
There is actually a language that is similar to Teal (Lua with types) and compiles down to C: https://nelua.io

In this space you can also check out Titan / Pallene and Terra.
